### PR TITLE
[HAL] NT6 compatibility HalpMapPhysicalMemory64 and HalpUnmapVirtualAddress

### PR DESCRIPTION
**Pull request by Getequ**

### CODE REVIEW

**Comment by ThFabba** [comment]
First of all, what's your goal here? Do you want our hal.dll to be usable on Vista?

My review feedback will change drastically depending on that... however if that's it, then:
I'd suggest making separate functions, e.g. HalpMapPhysicalMemory64Vista, have the regular HalpMapPhysicalMemory64 be a wrapper around the more powerful function, and then making only the KdMapPhysicalMemory64 assignment ifdef'd.

As a side note, there are no plans toward making an NT6 compatible kernel/HAL yet, so while I think that NTDDI_VERSION is probably not the right define to use for making the decision, I can't tell you what is.

Thanks!

**Comment by Getequ** [comment]
Goal is to make ReactOS buildable as NT6. Not to move from NT5 to NT6, just to make it possible. I see  there are so many places in a code ready for this and so many exe/dll's in cmake files have redefinition for _WIN32_WINNT and WINVER from 5.2 to 6.0 so i thought this is one of the minor intentions.

"HalpMapPhysicalMemory64 be a wrapper around the more powerful function" ok, fix it

**Comment by Getequ** [comment]
@ThFabba done.
my shiny goal is to make it possible to install .NET 4.5 in ReactOS. Maybe I will succeed in this if I do not abandon. Anyway it will be a little step forward for a project :)

File: hal/halx86/generic/memory.c [hunk 1]
```diff
@@ -199,14 +216,17 @@ HalpMapPhysicalMemory64(IN PHYSICAL_ADDRESS PhysicalAddress,
...
     return VirtualAddress;
 }
 
 VOID
 NTAPI
-HalpUnmapVirtualAddress(IN PVOID VirtualAddress,
-                        IN PFN_COUNT PageCount)
+HalpUnmapVirtualAddressVista(IN PVOID VirtualAddress,
+                        IN PFN_COUNT PageCount,
+                        IN BOOLEAN FlushCurrentTLB)
```
Comment (SergeGautherie) [inline hal/halx86/generic/memory.c]: Same here.

File: hal/halx86/generic/memory.c [hunk 2]
```diff
@@ -139,6 +139,23 @@ PVOID
...
+                        IN PFN_COUNT PageCount)
+{
+    HalpUnmapVirtualAddressVista(VirtualAddress, PageCount, TRUE);
+}
+
+PVOID
+NTAPI
+HalpMapPhysicalMemory64Vista(IN PHYSICAL_ADDRESS PhysicalAddress,
+                        IN PFN_COUNT PageCount,
+                        IN BOOLEAN FlushCurrentTLB)
```
Comment (SergeGautherie) [inline hal/halx86/generic/memory.c]: Please, align parameters.

**Comment by ThFabba** [comment]
I'm not opposed to this but I also don't fully understand the motivation. How is this a step towards .Net 4.5, which works in user mode and doesn't interact with HAL?

**Comment by Getequ** [comment]
one way is to change reporting version. another is to fix building as NT6 and reporting version. 

HAL is just the next issue that i met trying to build as winver=0x600 after fixing CORE-12596

**Comment by Getequ** [comment]
bussupp.c updated also

**Comment by SergeGautherie** [comment]
??install .NET 4.5??
Ftr, MS .NET Framework 4.5.x requires W. Vista SP2.

??change reporting version??
Fwiw, would a shim reporting ROS as NT 6.0 actually work enough?

**Comment by Getequ** [comment]
>4.5.x requires W. Vista SP2
Yes, and building as NT 6.0 is a first step. It is more correct way than just changing reporting version and anyway moving to NT6 unavoidable in a couple years i think. Wine upgraded to Win7 and even LibreOffice last version requires Win7 today.

>would a shim reporting ROS as NT 6.0 actually work enough?
No, installer detect compatibility mode and don't wanna start

>Ftr Fwiw
this abbreviations have mone then 3 meaning each :)

**Comment by learn-more** [comment]
@Getequ which setup did you try this with?

**Comment by Getequ** [comment]
@learn-more i used this https://www.microsoft.com/en-us/download/details.aspx?id=40779

**Comment by SergeGautherie** [comment]
> https://www.microsoft.com/en-us/download/details.aspx?id=40779

That is v4.5.1.
Any reason not to use [v4.5.2](https://www.microsoft.com/en-us/download/details.aspx?id=42642)?

**Comment by Getequ** [comment]
@SergeGautherie  i was looking for exactly 4.5 full installer, but found only web-installer, so next version i found is a 4.5.1. I think no difference what installer to use - both met the same problems.

Anyway - ROS has so many functions required for this installation but by some strange reason not exposed from their dll's. For example FlsAlloc/Free FlsGet/SetValue

File: hal/halx86/include/halp.h [hunk 3]
```diff
@@ -656,6 +672,7 @@ HalpUnmapVirtualAddress(
     IN PFN_COUNT NumberPages
 );
 
+
```
Comment (ThFabba) [inline hal/halx86/include/halp.h]: Remove this please

File: hal/halx86/generic/memory.c [hunk 4]
```diff
@@ -199,14 +216,17 @@ HalpMapPhysicalMemory64(IN PHYSICAL_ADDRESS PhysicalAddress,
     }
 
     /* Flush the TLB and return the address */
```
Comment (SergeGautherie) [inline hal/halx86/generic/memory.c]: Maybe split this comment in two, then?

**Pull request merged**